table/tables/tables.go

```diff
@@ -929,6 +939,12 @@ func (t *tableCommon) AllocAutoID(ctx sessionctx.Context) (int64, error) {
 	return rowID, nil
 }
 
+// overflowShardBits check whether the rowID overflow `1<<(64-t.meta.ShardRowIDBits) -1`.
+func (t *tableCommon) overflowShardBits(rowID int64) bool {
+	mask := uint64(^(1<<(64-t.meta.ShardRowIDBits) - 1))
+	return uint64(rowID)&mask > 0
```

Comment:
I think it should be `64-t.meta.ShardRowIDBits-1`, because in calcShard, the shard id takes `t.meta.ShardRowIDBits+1` bits: https://github.com/pingcap/tidb/pull/8936/files#diff-f92a6a9ee57cf92f8d63bd55d2edc05bR952

how about:
```go
upperBound := uint64(1<<(64-t.meta.ShardRowIDBits-1) - 1)
return uint64(rowID) > upperBound
```

Hope it would be more easy for understanding.

Reply:
I think bit operations is more readable.